lib/std/log.zig

```diff
@@ -0,0 +1,81 @@
+    /// has gone wrong or not, but the circumstances are worth investigating.
+    warn,
+    /// This log level is intended be used when an error has occurred but the
+    /// program is able to recover and continue running.
+    err,
+    /// This log level should only be used for irrecoverable errors as it will
+    /// trigger a panic after logging. Note that this will not log if the build
+    /// mode is release small but will still trigger a panic.
+    fatal,
+};
+
+/// The default log level is based on build mode. Note that in ReleaseSmall
+/// builds logging is totally disabled (log.fatal still causes a panic though).
+pub const default_level: Level = switch (builtin.mode) {
+    .Debug => .info,
+    .ReleaseSafe => .warn,
+    .ReleaseFast => .err,
+    .ReleaseSmall => .fatal,
+};
+
+/// The current log level. This is set to root.log_level if present, otherwise
+/// log.default_level.
+pub const level: Level = if (@hasDecl(root, "log_level")) root.log_level else default_level;
+
+fn log(comptime message_level: Level, comptime format: []const u8, args: var) void {
+    if (@enumToInt(message_level) < @enumToInt(level))
+        return;
+    if (@hasDecl(root, "log")) {
+        root.log(message_level, format, args);
+    } else {
+        // On release small builds we don't log because that would require storing
+        // the error message.
+        if (builtin.mode == .ReleaseSmall) return;
+        const held = debug.getStderrMutex().acquire();
+        defer held.release();
+        const stderr = debug.getStderrStream();
+        nosuspend stderr.print(format, args) catch return;
+    }
+}
+
+/// Log a message to stderr if log.level is info. This log level is intended to
+/// be used for any information about the state of the program that would be
+/// useful while debugging.
+pub fn info(comptime format: []const u8, args: var) void {
+    log(.info, format, args);
+}
+
+/// Log a message to stderr if log.level is info or warn. This log level is
+/// intended be used when it is uncertain if something has gone wrong or not,
+/// but the circumstances are worth investigating.
+pub fn warn(comptime format: []const u8, args: var) void {
+    log(.warn, format, args);
+}
+
+/// Log a message to stderr if log.level is info, warn, or err. This log
+/// level is intended be used when an error has occurred but the program is
+/// able to recover and continue running.
+pub fn err(comptime format: []const u8, args: var) void {
+    @setCold(true);
+    log(.err, format, args);
+}
+
+/// Log a message to stderr if log.level is info, warn, err, or fatal.
+/// This log level should only be used for irrecoverable errors as it will
+/// trigger a panic after logging. Note that this will not log if the build
+/// mode is release small but will still trigger a panic.
+pub fn fatal(comptime format: []const u8, args: var) noreturn {
+    @setCold(true);
+    log(.fatal, format, args);
+    @panic("");
```

Comment:
I don't think fatal logging should actually call `panic`: that should be up to the caller.

Reply:
I'm conflicted on this, so I took my cue from @andrewrk's example in #2586. I'm leaning towards removing the panic and leaving it up to the caller myself though.